syntax = "proto2";

message TimestampMessage {
    required uint64 id = 1;
    required uint64 timestamp = 2;
}

message ReadMessage {
    required string key = 1;
    required TimestampMessage readtime = 2;
}

message ReadReply {
     required string key = 1;
     required string value = 2;
     required uint64 timestamp = 3;
     required uint64 end = 4;
     required int32 op = 5;
}

message WriteMessage {
    required string key = 1;
    required string value = 2;
}

message IncrementMessage {
   required string key = 1;
   required string value = 2;
   required uint64 op = 3;
}


// A serialized transaction
message TransactionMessage {
    repeated ReadMessage readset = 1;
    repeated WriteMessage writeset = 2;
	repeated IncrementMessage incrementset = 3;
}
